### PR TITLE
Remove URL encoding of identifier

### LINKED ISSUE #1575: Identifiers url-encoded twice
I was wondering if Identifiers being url encoded twice is expected behavior ?
Right now it is encoded once by the IRI converter, and another one by the Symfony's route generation. Does it mean the client should always url-decode an @id before accessing it ?

### PR DESCRIPTION
**Pull request by silverbackdan**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tickets       |  fixes #1575
| License       | MIT
| Doc PR        | api-platform/docs#...

It looks as though the url encoding was not necessary and caused duplicate encoding in any case. The example I have tested originally returned `%253A` in place of the colon, which is the percent sign of %3A re-encoding.

This fix removes all URL encoding, it doesn't appear necessary in the IRI converter.

Would this class as a BC change instead of a fix though? Is it better to drill down further and still url encode the IRI once?

### CODE REVIEW

**Comment by silverbackdan** [comment]
Just to clarify that this should work well. The generated identifier is passed into `->router->generate` as a parameter. The Symfony router should handle url encoding any IDs as necessary.

I've clarified the use case a bit by including a character that should also be encoded in the `@id` key. Either way, it appears you can choose whether or not to encode any or the characters in this instance.

The Behat failing test appears to be unrelated. Please let me know if there's anything else though that I didn't spot.

File: src/JsonLd/Serializer/ItemNormalizer.php [hunk 1]
```diff
@@ -73,6 +73,7 @@ public function normalize($object, $format = null, array $context = [])
 
         $resourceClass = $this->resourceClassResolver->getResourceClass($object, $context['resource_class'] ?? null);
         $context = $this->initContext($resourceClass, $context);
+
```
Comment (soyuka) [inline src/JsonLd/Serializer/ItemNormalizer.php:76]: remove newline : 
```suggestion
```

File: features/main/url_encoded_id.feature [hunk 2]
```diff
@@ -0,0 +1,26 @@
+Feature: Allowing resource identifiers with characters that should be URL encoded
+  In order to have a resource with an id with special characters
+  As a client software developer
+  I need to be able to set and retrieve these resources with the URL encoded ID
+
+  @createSchema
+  Scenario Outline: Get a resource whether or not the id is URL encoded
+    Given there is a UrlEncodedId resource
+    And I add "Content-Type" header equal to "application/ld+json"
+    When I send a "GET" request to "<url>"
```
Comment (soyuka) [inline features/main/url_encoded_id.feature:10]: shouldn't this be `/url_encoded_ids/%encode:id` I don't really understand how `<url>` matches the entity

**Comment by dunglas** [comment]
I also fear that this change will introduce a BC break. We should take a deeper look to the URI RFC. It contains the list of characters that should and shouldn’t be encoded in the path section.

**Comment by dunglas** [comment]
For instance, Go has a `url.PathEscape()` as we’ll as a `url.QueryEscape()` function. They encode different classes of characters. AFAIU, we should use the equivalent of `PathEscape` here because chars such as `/` still have to be encoded.

File: features/main/url_encoded_id.feature (hunk 2 above)
Comment (silverbackdan) [reply features/main/url_encoded_id.feature:10]: It's from the scenario outline variable below. Line 21. So I'm testing many different ways of fetching the entity with and without encoding.

**Comment by dunglas** [comment]
If it fixes a double encoding issue, it looks good to me. We need to double check that this change doesn’t introduce regressions.

File: features/main/url_encoded_id.feature (hunk 2 above)
Comment (soyuka) [reply features/main/url_encoded_id.feature:10]: oh thanks I wasn't aware it was possible :)

**Comment by soyuka** [comment]
Would you be able to target master? It'll be safer to introduce this as a potentially breaking change :)

**Comment by silverbackdan** [comment]
@dunglas I'm not sure if it'd count as a regression, but the `@id` would no longer encode characters such as colons in the JSON output. However, you can still request the resource with the character encoded if you wanted to. The tests I've added go through variations of encoding the characters in the request to make sure the same resource is still returned. The slash is still encoded.

@soyuka sure no problem. thank you :) and you can thank Vincent for teaching me about Scenario Outlines ;)

**Comment by soyuka** [comment]
superseeded by #3611

**Pull request closed without merging**